src/index.js

```diff
@@ -0,0 +1,324 @@
+  }
+
+  startPolling() {
+    // try to acquire the master lock every 15-30 seconds by default
+    const masterMs = this.masterDuration * 500;
+    this.masterInterval = setInterval(() => {
+      const randomDelay = Math.floor(masterMs * Math.random());
+      setTimeout(() => this.acquireMasterLock(), randomDelay);
+    }, masterMs);
+
+    // re-register this worker every 10 seconds by default
+    const healthcheckMs = this.healthcheckDuration * 1000;
+    this.healthcheckInterval = setInterval(() => this.registerWorker(), healthcheckMs);
+  }
+
+  /**
+   * If the "master" lock is acquired, this worker will be responsible for
+   * assigning oplogs to workers
+   */
+  acquireMasterLock() {
+    const key = `${this.keyPrefix}:master`;
+    return this.lock.acquire(key, {
+      timeout: this.masterDuration * 1000,
+      retries: 0
+    })
+      .then(() => {
+        debug(`${this.workerId} is master.`);
+        return this.assignWorkers();
+      })
+      // ignore lock acquisition errors
+      .catch(err => {});
+  }
+
+  registerWorker() {
+    const key = this.workerRegistrationKey;
+    const now = Date.now();
+    return this.redisClient.zaddAsync(key, now, this.workerId)
+      .then(() => this.startNewAssignments())
+  }
+
+  getWorkerIds() {
+    const expiredTime = Date.now() - (this.healthcheckDuration * 1000);
+    const key = this.workerRegistrationKey;
+    return Promise.resolve()
+      // delete workers that have failed to renew their registration
+      .then(() => this.redisClient.zremrangebyscoreAsync(key, '-inf', expiredTime))
+      // get active workers
+      .then(() => this.redisClient.zrangeAsync(key, 0, -1));
+  }
+
+  getWorkerAssignmentsKey(workerId) {
+    return `${this.keyPrefix}:worker:${workerId}`;
+  }
+
+  readWorkerAssignmentsFromRedis(workerId) {
+    const key = this.getWorkerAssignmentsKey(workerId);
+    return this.redisClient.smembersAsync(key);
+  }
+
+  writeWorkerAssignmentsToRedis() {
+    return Promise.map(Object.keys(this.assignmentsByWorkerId), workerId => {
+      const key = this.getWorkerAssignmentsKey(workerId);
+      const connStrs = this.assignmentsByWorkerId[workerId];
+      return this.redisClient.saddAsync(key, connStrs)
+        .then(() => this.redisClient.expireAsync(key, this.masterDuration * 60));
+    });
+  }
+
+  recordAssignment(workerId, connStr) {
+    this.assignmentsByConnStr[connStr] = this.assignmentsByConnStr[connStr] || [];
+    this.assignmentsByConnStr[connStr].push(workerId);
+    this.assignmentsByWorkerId[workerId] = this.assignmentsByWorkerId[workerId] || [];
+    this.assignmentsByWorkerId[workerId].push(connStr);
+  }
+
+  getAvailableWorkerId() {
+    let availableWorkerId = this.workerId;
+    Object.keys(this.assignmentsByWorkerId).forEach(workerId => {
+      if (this.assignmentsByWorkerId[workerId].length < this.assignmentsByWorkerId[availableWorkerId]) {
+        availableWorkerId = workerId;
+      }
+    });
```

Comment:
This lookup finds the last `availableWorkerId` so it traverses through the entire list.
If using `Array.prototype.find`, then it will exit after the first valid `workerId`

Reply:
It finds the worker with the least number of assignments, so it must traverse the entire list.

Reply:
Makes sense. Here is food for thought:
If `assignmentsByWorkerId` is a big list, doing this loop could be inefficient. It might be more performant maintain an ordered list when calling `recordAssignment` and then fetching the head or tail of the list.

Reply:
Good thinking, but I'd prefer to choose simplicity over performance here. The list will never be huge so the extra code will have a negligible performance gain.